### PR TITLE
[macOS] Forward key events to NSTextInputContext when there's composing text

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Partially fixes https://github.com/flutter/flutter/issues/85328, still does not work in the macOS accent menu

Quick demo:

https://user-images.githubusercontent.com/31859944/158485133-ef552aeb-022f-491d-9753-b4beb6298d84.mov

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm [hunk 1]
```diff
@@ -356,6 +364,40 @@ - (bool)textInputPlugin {
...
+  [tester.manager handleEvent:keyDownEvent(0x50)];
+
+  NSMutableArray<FlutterAsyncKeyCallback>* channelCallbacks =
+      [NSMutableArray<FlutterAsyncKeyCallback> array];
+  NSMutableArray<FlutterAsyncKeyCallback>* embedderCallbacks =
+      [NSMutableArray<FlutterAsyncKeyCallback> array];
+  [tester recordEmbedderCallsTo:embedderCallbacks];
+  [tester recordChannelCallsTo:channelCallbacks];
+  // TextInputPlugin does not claim the event.
+  [tester respondTextInputWith:NO];
```
Comment (dkwingsmt) [inline shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]: This method is used to set the responses from the mocked TextInputPlugin, not to check. (Maybe I should add a method to check its value...)

Comment (dkwingsmt) [inline shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]:
```suggestion
```

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm [hunk 2]
```diff
@@ -356,6 +364,40 @@ - (bool)textInputPlugin {
...
+  [tester.manager handleEvent:keyUpEvent(0x50)];
+
+  // Nobody gets the event except for the text input plugin.
+  EXPECT_EQ([channelCallbacks count], 0u);
+  EXPECT_EQ([embedderCallbacks count], 0u);
+
+  // Send another down event with composing == NO.
+  tester.isComposing = NO;
+  [tester respondTextInputWith:YES];
```
Comment (dkwingsmt) [inline shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:390]:
```suggestion
```

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm (hunk 1 above)
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]: Or can you create a way for the mocked text input plugin to record the events, just like the embedder and channel ones?

**Review by dkwingsmt — CHANGES_REQUESTED** [review]

**Comment by dkwingsmt** [comment]
Also, why is this not a long-term fix? I thought this was the only way to fix it.

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm (hunk 1 above)
Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]: Yeah I didn't mean to use this as a "verification" statement. The verification step is at line 385 & 386 where we make sure the event doesn't go to the primary responder(s) even with text input not claiming the key event for itself.

Comment (LongCatIsLooong) [reply shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]: sure ideally I want to verify that the text input plugin did get the event.

Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterKeyboardManagerUnittests.mm:381]: Yeah so this line does little and should be removed. And still we haven't ensured that `onTextInputKeyEvent` is called.

**Comment by LongCatIsLooong** [comment]
> Also, why is this not a long-term fix? I thought this was the only way to fix it.

macOS has all kinds of [popover](https://developer.apple.com/design/human-interface-guidelines/macos/windows-and-views/popovers/)s that can be displayed during text input and should receive key events when visible, the IME candidate window is just one of them (other examples, the accent menu or the [emoji menu](https://www.macrumors.com/how-to/use-emoji-on-mac/)). `!isComposing` doesn't mean there's no popover.

**Review by dkwingsmt — APPROVED** [review]
LGTM

**Pull request merged**